Add ability to compute HOG probabilities for a given sampler

KevinVillela opened the pull request:
#2111 
This also adds a Jupyter notebook walk-through of the algorithm. The next and final step here would be to actually compile the circuit to our gate set(which I believe @john6060 is working on), route it (using contrib/routing), and then optimize it (with something in the optimizers/ directory).

Strilanc commented on examples/advanced/quantum_volume.py:
```diff
@@ -104,6 +104,41 @@ def compute_heavy_set(circuit: cirq.Circuit) -> List[int]:
+
+    """
+    # Add measure gates to the end of the circpuit.
+    circuit_copy = circuit.copy()  # So we don't mutate the original circuit.
```
`circuit + cirq.measure(...)`

KevinVillela replied:
Done, although TBH I'm not sure it's cleaner :\

Strilanc replied:
Oh right, adding the operation directly still doesn't work. Bah. Whichever.

Strilanc replied:
Use `cirq.Circuit(` instead of `cirq.Circuit.from_ops(`

KevinVillela replied:
Ooh actually you can add a list of gates to the circuit, I wrapped it in brackets and it worked :) Thanks!

Strilanc replied:
After https://github.com/quantumlib/Cirq/pull/2189 merges you can sync with master and add the operation without the wrapping list.

KevinVillela replied:
Sweet, thanks! Synced.

Strilanc commented on examples/advanced/quantum_volume.py:
```diff
@@ -104,6 +104,41 @@ def compute_heavy_set(circuit: cirq.Circuit) -> List[int]:
+
+    # Compute the number of outputs that are in the heavy set.
+    num_in_heavy_set = 0
+    for _, m in enumerate(measurements.data.iterrows()):
```
You're enumerating to add the index, then ignoring the index. Just iterate without enumerate.

KevinVillela replied:
Woops, nice catch. Done.

Strilanc commented on examples/advanced/quantum_volume.ipynb:
```diff
@@ -0,0 +1,254 @@
+    "\n",
+    "for depth in depths:\n",
+    "    num_qubits = depth # Square matrix.\n",
+    "    print(\"Running simulation with %s qubits and a depth of %s\" % (num_qubits, depth))\n",
```
Prefer fstrings. This applies in a lot of places.

KevinVillela replied:
Oh good to know, done.

Strilanc commented on examples/advanced/quantum_volume.ipynb:
```diff
@@ -0,0 +1,254 @@
+    "    average_ideals.append(average_ideal)\n",
+    "    average_noisies.append(average_noisy)\n",
+    "\n",
+    "# Create the chart. Note that we try to make it look as similar\n",
```
Might be clearer if extracted into a method.

Is the call to `plt.show()` required?

KevinVillela replied:
Done, and no it was not, so I removed.

KevinVillela commented:
Thanks for the quick review, I've addressed all the comments.
(Btw I'm not sure if it's customary in Github to make a comment here after each round of review, let me know if that isn't normal. I also don't know if I should be resolving comment threads.)

KevinVillela commented:
Thanks for the 3 AM review :P

mpharrigan commented on examples/advanced/quantum_volume.py:
```diff
@@ -104,6 +104,40 @@ def compute_heavy_set(circuit: cirq.Circuit) -> List[int]:
+    circuit_copy = circuit + [cirq.measure(*sorted(circuit.all_qubits()))]
+
+    # Run the sampler to compare each output against the Heavy Set.
+    repetitions = 1000
```
this should likely be an argument to the function, perhaps with this default

KevinVillela replied:
Done

mpharrigan commented on examples/advanced/quantum_volume.py:
```diff
@@ -104,6 +104,40 @@ def compute_heavy_set(circuit: cirq.Circuit) -> List[int]:
+
+    # Compute the number of outputs that are in the heavy set.
+    num_in_heavy_set = 0
+    for m in measurements.data.iterrows():
```
You can try to use `np.in1d` (and `np.sum`) to vectorize this, which might give you noticeable speed increases at high `repetitions` number

KevinVillela replied:
Holy crap so much faster. Thanks!

mpharrigan commented on examples/advanced/quantum_volume.ipynb:
```diff
@@ -0,0 +1,109 @@
+   "outputs": [],
+   "source": [
+    "from examples.advanced import quantum_volume\n",
+    "import numpy as np\n",
```
import cirq

KevinVillela replied:
Done.

mpharrigan commented on examples/advanced/quantum_volume.ipynb:
```diff
@@ -0,0 +1,109 @@
+{
+ "cells": [
```
I'd split this up into >1 cells to be less intimidating for the reader

KevinVillela replied:
Done.

mpharrigan commented on examples/advanced/quantum_volume.ipynb:
```diff
@@ -0,0 +1,109 @@
+    "        print(f\"        Ideal simulation HOG probability: {ideal}\")\n",
+    "\n",
+    "        # Run a noisy simulator over the model circuit.\n",
+    "        noisy = quantum_volume.sample_heavy_set(model_circuit, heavy_set, sampler=cirq.DensityMatrixSimulator(noise=cirq.ConstantQubitNoiseModel(qubit_noise_gate=cirq.DepolarizingChannel(p=0.005))))\n",
```
construct the `sampler` outside of the loop so the user can tinker with it; slot in different samplers; etc. This would also make this line less long :)

KevinVillela replied:
Done

mpharrigan commented on examples/advanced/quantum_volume.ipynb:
```diff
@@ -0,0 +1,109 @@
+    "    average_noisies.append(average_noisy)\n",
+    "\n",
+    "# Create the chart.\n",
+    "generate_chart(depths, average_ideals, average_noisies)"
```
Yeah I'd put `generate_chart` in its own cell so that you can collapse the printed output from the actual run and then have the plot show up without having to scroll so far

KevinVillela replied:
Done

mpharrigan commented:
Cool! My comments are optional style things

mpharrigan commented on examples/advanced/quantum_volume.ipynb:
```diff
@@ -0,0 +1,109 @@
+    "    axs.scatter([d + .1 for d in depths], average_noisies, c='tab:red', label='Noisy')\n",
+    "    # Line markers for asymptotic ideal heavy output probability and the ideal Heavy Output Generation threshold.\n",
+    "    axs.hlines((1 + np.log(2)) / 2, 0, 100, colors=['tab:green'], label='Asymptotic ideal', linestyles='dashed')\n",
+    "    axs.hlines(2/3, 0, 100, label='HOG threshold', linestyles='dotted')\n",
```
use `axs.axhline`. Then you can remove `(0, 100)`. Then you can remove `set_xlim`. Then the code works if you change the depths to include values > 4

KevinVillela replied:
Awesome, that's super helpful! Done.

KevinVillela commented:
Very helpful comments, I'm learning a lot, thanks! PTAL

Strilanc commented:
We don't have an official flow for how a review should go. I usually delegate resolving the comment to the person who was asking for something to be done.

It's all made kinda moot by Github being terrible at showing resolved comments in the context of the resolution.

Strilanc commented on examples/advanced/quantum_volume.ipynb:
```diff
@@ -0,0 +1,127 @@
+    "\n",
+    "# Here is the important set-up: The samplers. These are what will be\n",
+    "# run on the generated circuit, and then evaluated.\n",
+    "class SamplerConfig:\n",
```
This class is mis-named. You're mostly configuring the plot as opposed to the sampler.

The class is so small, and the context so local, that you might just want to use a NamedTuple or even just tuple. You can merge all the scatter plot stuff into a single kwargs, leaving you with `Tuple[cirq.Sampler, Dict[str, Any]]`.

Anything you can do to simplify this code, do it.

KevinVillela replied:
Sounds good, I made it a simple Dict[str, Any] for now, let me know if you prefer me to separate the Sampler and the plotting config.

Strilanc commented on examples/advanced/quantum_volume.ipynb:
```diff
@@ -0,0 +1,127 @@
+   "outputs": [],
+   "source": [
+    "from matplotlib import pyplot as plt\n",
+    "from matplotlib.ticker import FormatStrFormatter\n",
```
Unused import

KevinVillela replied:
Done.

Strilanc commented on examples/advanced/quantum_volume.ipynb:
```diff
@@ -0,0 +1,127 @@
+    "            self.color = color\n",
+    "            self.marker = marker      \n",
+    "            # Keep track of the probabilities.\n",
+    "            self.probabilities = [0] * len(depths)\n",
```
The existence of this field directly contradicts the notion that this is a config.

KevinVillela replied:
Changed this a bit.

Strilanc commented on examples/advanced/quantum_volume.ipynb:
```diff
@@ -0,0 +1,127 @@
+    "    # Compute the average performance over the total number of runs.\n",
+    "    for config in samplers:\n",
+    "        config.probabilities[depth - len(depths)] /= num_repetitions\n",
+    "        print(f\"    Average {config.label} HOG probability: {config.probabilities[depth - len(depths)]}\")\n"
```
On my screen in the github preview these strings are long enough that they wrap. Reflow long lines.

KevinVillela replied:
Ran this through yapf, should look better now.

mpharrigan commented:
Not a huge fan of the new config class in the notebook. It's ok to RYS (repeat yourself sometimes; my trademark comeback to DRY) in notebooks to make it easier to read / tinker / extend

KevinVillela commented:
Made some more changes to the notebook to simplify it, and added a top config cell.
